src/abi/abi_type.ts

```diff
@@ -0,0 +1,846 @@
+      const seg = dynamicSegments[i];
+      if (seg.left > seg.right) {
+        throw new Error(
+          'dynamic segment should display a [l, r] space with l <= r'
+        );
+      }
+      if (
+        i !== dynamicSegments.length - 1 &&
+        seg.right !== dynamicSegments[i + 1].left
+      ) {
+        throw new Error('dynamic segment should be consecutive');
+      }
+    }
+
+    // Check dynamic element partitions
+    let segIndex = 0;
+    // eslint-disable-next-line no-shadow
+    for (let i = 0; i < tupleTypes.length; i++) {
+      if (tupleTypes[i].IsDynamic()) {
+        valuePartition[i] = byteString.slice(
+          dynamicSegments[segIndex].left,
+          dynamicSegments[segIndex].right
+        );
+        segIndex += 1;
+      }
+    }
+
+    // Decode each tuple element
+    const returnValues: ABIValue[] = [];
+    // eslint-disable-next-line no-shadow
+    for (let i = 0; i < tupleTypes.length; i++) {
+      const valueTi = tupleTypes[i].Decode(valuePartition[i]);
+      returnValues.push(valueTi);
+    }
+    return returnValues;
+  }
+
+  // Assume that the current index on the list of type is an ABI bool type.
+  // It returns the difference between the current index and the index of the furthest consecutive Bool type.
+  static findBoolLR(typeList: Type[], index: number, delta: -1 | 1): number {
+    let until = 0;
+    while (true) {
+      const curr = index + delta * until;
+      if (typeList[curr].constructor === BoolType) {
+        if (curr !== typeList.length - 1 && delta === 1) {
+          until += 1;
+        } else if (curr > 0 && delta === -1) {
+          until += 1;
+        } else {
+          break;
+        }
+      } else {
+        until -= 1;
+        break;
+      }
+    }
+    return until;
+  }
+
+  static parseTupleContent(str: String): String[] {
+    if (str.length === 0) {
+      return [];
+    }
+    if (str.endsWith(',') || str.startsWith(',')) {
+      throw new Error('tuple string should not start with comma');
+    }
+    if (str.includes(',,')) {
+      throw new Error('tuple string should not have consecutive commas');
+    }
+
+    const tupleStrings: string[] = [];
+    let depth = 0;
+    let word = '';
+
+    for (const char of str) {
+      word += char;
+      if (char === '(') {
+        depth += 1;
+      } else if (char === ')') {
+        depth -= 1;
+      } else if (char === ',') {
+        // If the comma is at depth 0, then append the word as token.
+        if (depth === 0) {
+          tupleStrings.push(word.slice(0, word.length - 1));
+          word = '';
+        }
+      }
+    }
+    if (word.length !== 0) {
+      tupleStrings.push(word);
+    }
+    if (depth !== 0) {
+      throw new Error('tuple string has mismatched parentheses');
+    }
+    return tupleStrings;
+  }
+
+  // compressMultipleBool compresses consecutive bool values into a byte in ABI tuple / array value.
+  static compressMultipleBool(valueList: ABIValue[]): number {
```

Comment:
nit: functions `compressMultipleBool` and `findBoolLR` aren't really related to `TupleType`, so I'd suggest making them their own standalone functions instead of static functions on that class.